dozer/cogs/roles.py

```diff
@@ -68,6 +76,25 @@ class Roles(Cog):
             if dbrole is not None:
                 session.delete(dbrole)
 
+    async def ctxpurge(self, ctx, returncounter):
+        counter = 0
+        with db.Session() as session:
+            roles = session.query(GiveableRole).filter_by(guild_id=ctx.guild.id)
+            guildroles = []
+            for i in ctx.guild.roles:
+                guildroles.append(i.id)
+            for role in roles:
+                if role.id not in guildroles:
+                    await self.givemepurge(role)
```

Comment:
This still creates a new transaction for every role deleted, when deleting all missing roles could be one query. Given a list of all the role IDs in a guild, you can find records for that guild whose role ID is [not in](http://docs.sqlalchemy.org/en/latest/core/sqlelement.html#sqlalchemy.sql.expression.ColumnElement.notin_) that collection, then [delete](http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.delete) them all at once.